### PR TITLE
Optimizations related to BlockEditorPropertyEditor.ToEditor()

### PR DESCRIPTION
**Pull request by elit0451**
Aims at eliminating/decreasing the Block editor performance issues.
Reference: https://github.com/umbraco/Umbraco-CMS/issues/10184

## Details
<s>- Changing the built-in cache at **BlockEditorValues** level for Content Types - before that dictionary cache was never used because it was short-lived and re-created while looping through block items. Now, it is created in **BlockEditorPropertyEditor** and passed on, so it can be reused.</s>
- Calling `GetValueEditor()` for unique data type ids - reducing the number of objects allocated
- Adding explicit scope around `GetEmptyByKey()` - so everything is handled in 1 scope
- Introducing `GetEmptyByKeys()` which collects all the block list editor items and executes a single call for block list editor instead of making numerous requests as mentioned in #10184 
- Fix tests
</br>

_Before_:

⬆️ all ❗ marks from the marked one down are all from GetEmptyByKey  

</br>
</br>
</br>

_After_:
image

## Test
- Follow the steps mentioned in the above issue and review the optimizations

---
Note:
This PR contains client-side code changes, so remember to run `npm run build` in Umbraco.Web.UI.Client

### CODE REVIEW

File: src/Umbraco.Web/Editors/ContentController.cs [hunk 1]
```diff
@@ -367,13 +373,17 @@ public ContentItemDisplay GetEmpty(string contentTypeAlias, int parentId)
...
-            }
+                var contentType = Services.ContentTypeService.Get(contentTypeKey);
+                if (contentType == null)
+                {
+                    throw new HttpResponseException(HttpStatusCode.NotFound);
+                }
+                scope.Complete();
 
-            return GetEmpty(contentType, parentId);
+                return GetEmpty(contentType, parentId);
```
Comment (nikolajlauridsen) [inline src/Umbraco.Web/Editors/ContentController.cs:385]: I could be wrong, but I think we should call `scope.Complete()` after doing `GetEmpty` as far as I can see `GetEmpty` performs write operations through `Services.ContentService.Create` so if something goes wrong I think we want to have that rolled back.

**Review by nikolajlauridsen — COMMENTED** [review]
Overall looks good to me, found one minor thing in ContentController though

File: src/Umbraco.Web/Editors/ContentController.cs (hunk 1 above)
Comment (elit0451) [reply src/Umbraco.Web/Editors/ContentController.cs:385]: Good point!

**Review by nikolajlauridsen — APPROVED** [review]
Just finished testing, everything seems to work and I'm seeing a performance improvement as well, as you also showed in your description 👍

File: src/Umbraco.Web/Editors/ContentController.cs [hunk 2]
```diff
@@ -367,13 +373,19 @@ public ContentItemDisplay GetEmpty(string contentTypeAlias, int parentId)
...
-            }
+                var contentType = Services.ContentTypeService.Get(contentTypeKey);
+                if (contentType == null)
+                {
+                    throw new HttpResponseException(HttpStatusCode.NotFound);
+                }
 
-            return GetEmpty(contentType, parentId);
+                var contentItem = GetEmpty(contentType, parentId);
+                scope.Complete();
```
[the remark this answers is not in the document]
Comment (elit0451) [reply src/Umbraco.Web/Editors/ContentController.cs:385]: Good tip, thanks, will have it in mind for the next time ;)

File: src/Umbraco.Web/Editors/ContentController.cs [hunk 3]
```diff
@@ -392,6 +404,37 @@ private ContentItemDisplay GetEmpty(IContentType contentType, int parentId)
...
+                var contentTypes = Services.ContentTypeService.GetAll(contentTypeKeys).ToList();
+
+                if (contentTypes.Any(contentType => contentType == null))
+                {
+                    throw new HttpResponseException(HttpStatusCode.NotFound);
+                }
+
+                foreach (var contentTypeKey in contentTypeKeys)
+                {
+                    var contentType = contentTypes.First(c => c.Key == contentTypeKey);
```
[the remark this answers is not in the document]
Comment (elit0451) [reply src/Umbraco.Web/Editors/ContentController.cs:428]: You are absolutely right, a total oversight and misleading from the IDE static optimizer. I Will fix it up, thanks 🙌

File: src/Umbraco.Web/Editors/ContentController.cs [hunk 4]
```diff
@@ -392,6 +404,32 @@ private ContentItemDisplay GetEmpty(IContentType contentType, int parentId)
...
+        public IDictionary<Guid, ContentItemDisplay> GetEmptyByKeys([FromUri] Guid[] contentTypeKeys, [FromUri] int parentId)
+        {
+            var result = new Dictionary<Guid, ContentItemDisplay>();
+
+            using var scope = _scopeProvider.CreateScope(autoComplete: true);
+            var contentTypes = Services.ContentTypeService.GetAll(contentTypeKeys).ToList();
+
+            foreach (var contentType in contentTypes)
+            {
+                if (contentType is null)
```
Comment (Shazwazza) [inline src/Umbraco.Web/Editors/ContentController.cs:422]: This whole block could just be:

```cs
var result = Services.ContentTypeService.GetAll(contentTypeKeys)
   .WhereNotNull()
   .ToDictionary(x => x.Key, x => GetEmpty(x, parentId));
```

File: src/Umbraco.Web/PropertyEditors/BlockEditorPropertyEditor.cs [hunk 5]
```diff
@@ -118,7 +120,7 @@ public override object ToEditor(Property property, IDataTypeService dataTypeServ
                     return string.Empty;
                 }
 
-                if (blockEditorData == null || blockEditorData.BlockValue.ContentData.Count == 0)
+                if (blockEditorData == null)
```
Comment (Shazwazza) [inline src/Umbraco.Web/PropertyEditors/BlockEditorPropertyEditor.cs:123]: do you recall why this change was made?

**Review by Shazwazza — COMMENTED** [review]
@elit0451 + @nikolajlauridsen Left a question and a minor comment. I don't think these should block merging this PR but if you can respond (or make the change) that would be great. @nikolajlauridsen I'll look at your other PR now but if you have time tomorrow, feel free to merge this one and we'll flag as 8.15.

File: src/Umbraco.Web/PropertyEditors/BlockEditorPropertyEditor.cs (hunk 5 above)
Comment (elit0451) [reply src/Umbraco.Web/PropertyEditors/BlockEditorPropertyEditor.cs:123]: If you have a closer look to what exactly `blockEditorData = _blockEditorValues.DeserializeAndClean(val)` is doing (a few lines above), you will notice that we already do the check I removed ([`blockEditorData.BlockValue.ContentData.Count == 0`](https://github.com/umbraco/Umbraco-CMS/blob/d8b7e24b7f575e8b756f950e4de43a21e4da8795/src/Umbraco.Web/PropertyEditors/BlockEditorPropertyEditor.cs#L353)). And if the count is indeed 0, we will return `null` that would be caught in the single `is null` check 🙂

Comment (nikolajlauridsen) [reply src/Umbraco.Web/PropertyEditors/BlockEditorPropertyEditor.cs:123]: Ahh yes, and `DeserializeAndClean` might return null, hence the null check 👍

File: src/Umbraco.Web/Editors/ContentController.cs (hunk 4 above)
Comment (nikolajlauridsen) [reply src/Umbraco.Web/Editors/ContentController.cs:422]: Ah good catch, however, I'm changing this method in the other PR, so I think I'll just merge this as-is for now, just to avoid a merge conflict 😄
